File: examples/nodejs/credential_plugin.js
#!/usr/bin/env node
/**
 * Executa 凭据插件示例 — Node.js 实现
 *
 * 演示如何：
 * 1. 在 Manifest 中声明所需凭据（credentials 字段）
 * 2. 在 invoke 中从 context.credentials 读取凭据
 * 3. 回退到环境变量以支持本地开发
 * 4. 安全地使用凭据调用外部 API
 *
 * 凭据的三层解析优先级：
 *   1. 平台统一凭据 — 用户在 /settings/authorizations 一次性配置
 *   2. 插件级凭据   — 用户在单个插件设置中手动填写
 *   3. 环境变量     — 本地开发时从 process.env 读取（插件自行实现）
 *
 * Agent 会将解析后的凭据通过 invoke 请求的 params.context.credentials 注入，
 * LLM 不会看到凭据内容，也无法在对话中泄露。
 *
 * 运行方式：
 *   node credential_plugin.js
 *
 * 本地开发（通过环境变量提供凭据）：
 *   GITHUB_TOKEN=ghp_xxx node credential_plugin.js
 *
 * 协议要求：
 *   - stdin:  接收 JSON-RPC 请求（每行一个 JSON 对象）
 *   - stdout: 返回 JSON-RPC 响应（每行一个 JSON 对象）
 *   - stderr: 日志输出（不会干扰协议通信）
 */

const readline = require("readline");

// ─── Manifest（自描述清单） ──────────────────────────────────────────
//
// credentials: 声明本插件所需的凭据列表
//
// credentials[].name 命名最佳实践：
//   - 使用全大写蛇形命名（如 GITHUB_TOKEN）
//   - 与平台提供商的 credential_mapping 对齐，实现自动映射
//     例如：GITHUB_TOKEN、TWITTER_API_KEY、GOOGLE_ACCESS_TOKEN
//   - 自定义服务用 SERVICE_NAME + 字段类型 命名
//
// sensitive=true 的凭据会在 UI 中以密码框显示，不回显明文。

const MANIFEST = {
  name: "github-tool",
  display_name: "GitHub Tool",
  version: "1.0.0",
  description: "GitHub 仓库查询工具，演示凭据（API Token）的声明与使用",
  author: "Anna Developer",
  credentials: [
    {
      name: "GITHUB_TOKEN",
      display_name: "Personal Access Token",
      description:
        "GitHub Settings → Developer Settings → Personal access tokens (fine-grained recommended)",
      required: true,
      sensitive: true,
    },
  ],
  tools: [
    {
      name: "get_repo",
      description: "查询 GitHub 仓库信息",
      parameters: [
        {
          name: "owner",
          type: "string",
          description: "仓库所有者（用户名或组织名）",
          required: true,
        },
        {
          name: "repo",
          type: "string",
          description: "仓库名称",
          required: true,
        },
      ],
    },
    {
      name: "list_issues",
      description: "列出仓库的 Issues",
      parameters: [
        {
          name: "owner",
          type: "string",
          description: "仓库所有者",
          required: true,
        },
        {
          name: "repo",
          type: "string",
          description: "仓库名称",
          required: true,
        },
        {
          name: "state",
          type: "string",
          description: "Issue 状态: open / closed / all（默认 open）",
          required: false,
          default: "open",
        },
        {
          name: "limit",
          type: "integer",
          description: "返回数量上限（1-30，默认 10）",
          required: false,
          default: 10,
        },
      ],
    },
  ],
  runtime: {
    type: "npm",
    min_version: "1.0.0",
  },
};

// ─── 凭据读取辅助函数 ─────────────────────────────────────────────
//
// 最佳实践：优先从 context.credentials 读取，回退到环境变量

/**
 * 获取凭据值（按优先级解析）
 * @param {Object|null} credentials - invoke 注入的 context.credentials
 * @param {string} name - 凭据名称
 * @param {string} [defaultValue] - 默认值
 * @returns {string|undefined}
 */
function getCredential(credentials, name, defaultValue) {
  const creds = credentials || {};
  return creds[name] || process.env[name] || defaultValue;
}

// ─── 工具实现 ─────────────────────────────────────────────────────

function toolGetRepo(args, credentials) {
  const { owner, repo } = args;
  const token = getCredential(credentials, "GITHUB_TOKEN");

  if (!token) {
    return {
      error: "GITHUB_TOKEN not configured",
      hint: [
        "配置方式（任选其一）：",
        "  1. 平台统一授权: /settings/authorizations 页面配置",
        "  2. 插件级凭据: Anna Admin → 插件设置 → 凭据配置",
        "  3. 本地开发: GITHUB_TOKEN=ghp_xxx node credential_plugin.js",
      ].join("\n"),
    };
  }

  // ─── 实际调用示例（注释） ───
  // const https = require("https");
  // const options = {
  //   hostname: "api.github.com",
  //   path: `/repos/${owner}/${repo}`,
  //   headers: {
  //     "Authorization": `Bearer ${token}`,
  //     "User-Agent": "anna-executa-plugin",
  //     "Accept": "application/vnd.github+json",
  //   },
  // };
  // ────────────────────────────

  // 模拟数据（演示用）
  return {
    full_name: `${owner}/${repo}`,
    description: "A simulated repository for demonstration",
    stars: 42,
    forks: 7,
    language: "TypeScript",
    open_issues: 3,
    token_configured: true,
    token_preview:
      token.length > 8 ? `${token.slice(0, 4)}...${token.slice(-4)}` : "***",
    _note: "This is simulated data for demonstration purposes",
  };
}

function toolListIssues(args, credentials) {
  const { owner, repo, state = "open", limit = 10 } = args;
  const token = getCredential(credentials, "GITHUB_TOKEN");

  if (!token) {
    return {
      error: "GITHUB_TOKEN not configured",
      hint: [
        "配置方式（任选其一）：",
        "  1. 平台统一授权: /settings/authorizations 页面配置",
        "  2. 插件级凭据: Anna Admin → 插件设置 → 凭据配置",
        "  3. 本地开发: GITHUB_TOKEN=ghp_xxx node credential_plugin.js",
      ].join("\n"),
    };
  }

  const effectiveLimit = Math.max(1, Math.min(30, limit));

  // 模拟数据
  const issues = [];
  for (let i = 0; i < effectiveLimit; i++) {
    issues.push({
      number: i + 1,
      title: `Example issue #${i + 1}`,
      state: state === "all" ? (i % 2 === 0 ? "open" : "closed") : state,
      labels: i % 3 === 0 ? ["bug"] : ["enhancement"],
      created_at: new Date(Date.now() - i * 86400000).toISOString(),
    });
  }

  return {
    repository: `${owner}/${repo}`,
    state,
    count: issues.length,
    issues,
    token_configured: true,
    _note: "This is simulated data for demonstration purposes",
  };
}

const TOOL_DISPATCH = {
  get_repo: toolGetRepo,
  list_issues: toolListIssues,
};

// ─── JSON-RPC 处理 ───────────────────────────────────────────────

function makeResponse(id, result, error) {
  const resp = { jsonrpc: "2.0", id };
  if (error !== undefined) resp.error = error;
  else resp.result = result;
  return resp;
}

function handleRequest(line) {
  let request;
  try {
    request = JSON.parse(line);
  } catch {
    return makeResponse(null, undefined, {
      code: -32700,
      message: "Parse error",
    });
  }

  const { id, method, params = {} } = request;

  switch (method) {
    case "describe":
      return makeResponse(id, MANIFEST);

    case "invoke": {
      const { tool, arguments: args = {}, context = {} } = params;
      const credentials = context.credentials || null;

      if (!tool) {
        return makeResponse(id, undefined, {
          code: -32602,
          message: "Missing 'tool' in params",
        });
      }
      const fn = TOOL_DISPATCH[tool];
      if (!fn) {
        return makeResponse(id, undefined, {
          code: -32601,
          message: `Unknown tool: ${tool}`,
          data: { available_tools: Object.keys(TOOL_DISPATCH) },
        });
      }
      try {
        // 将凭据作为第二个参数传入工具函数
        const result = fn(args, credentials);
        return makeResponse(id, { success: true, data: result, tool });
      } catch (e) {
        return makeResponse(id, undefined, {
          code: -32603,
          message: `Tool execution failed: ${e.message}`,
        });
      }
    }

    case "health":
      return makeResponse(id, {
        status: "healthy",
        timestamp: new Date().toISOString(),
        version: MANIFEST.version,
        tools_count: MANIFEST.tools.length,
        credentials_declared: MANIFEST.credentials.length,
      });

    default:
      return makeResponse(id, undefined, {
        code: -32601,
        message: `Method not found: ${method}`,
      });
  }
}

// ─── 主循环（stdio JSON-RPC 服务） ──────────────────────────────

const rl = readline.createInterface({ input: process.stdin });

process.stderr.write("🔌 GitHub credential plugin started (Node.js)\n");
process.stderr.write(
  `   Tools: ${Object.keys(TOOL_DISPATCH).join(", ")}\n`
);
process.stderr.write(
  `   Credentials required: ${MANIFEST.credentials.map((c) => c.name).join(", ")}\n`
);

rl.on("line", (line) => {
  line = line.trim();
  if (!line) return;

  process.stderr.write(`← ${line}\n`);
  const responseObj = handleRequest(line);
  const payload = JSON.stringify(responseObj);
  process.stdout.write(payload + "\n");
  process.stderr.write(`→ ${payload}\n`);
});

File: examples/python/credential_plugin.py
#!/usr/bin/env python3
"""Executa Credential Plugin Example — Demonstrates how to use credentials (API Key / Token)

This example shows how to:
1. Declare required credentials in the Manifest (credentials field)
2. Read credentials from context.credentials in invoke
3. Securely use credentials to call external APIs
4. Fall back to environment variables for local development

Credential resolution priority (three tiers):
  1. Platform unified credentials — configured once at /settings/authorizations
  2. Plugin-level credentials — manually entered in per-plugin settings
  3. Environment variables — read from os.environ for local development (plugin implements)

The Agent injects resolved credentials via invoke request's params.context.credentials.
LLM cannot see credential values and cannot leak them in conversations.

Usage:
    python credential_plugin.py

Local development (provide credentials via environment variables):
    WEATHER_API_KEY=your_key python credential_plugin.py

Protocol requirements:
    - stdin:  Receives JSON-RPC requests (one JSON object per line)
    - stdout: Returns JSON-RPC responses (one JSON object per line)
    - stderr: Log output (does not interfere with protocol communication)
"""

import json
import os
import sys
from datetime import datetime, timezone


# ─── Manifest (Self-Description) ─────────────────────────────────────
#
# credentials: Declares the list of credentials required by this plugin
#   - name:         Credential identifier (the key used when passed in, e.g. WEATHER_API_KEY)
#   - display_name: Human-readable name (displayed in the UI)
#   - description:  Usage description (helps users understand the credential's purpose)
#   - required:     Whether configuration is required (tool may not work properly if missing)
#   - sensitive:    Whether it is sensitive data (when true, Nexus encrypts storage; UI does not echo)

MANIFEST = {
    "name": "weather-tool",
    "display_name": "Weather Tool",
    "version": "1.0.0",
    "description": "Weather query tool demonstrating credential (API Key) declaration and usage",
    "author": "Anna Developer",
    # ─── Credential Declaration ───────────────────────────────────────
    # credentials[].name is the unique identifier; Agent uses it as the key for injection.
    #
    # Naming best practices:
    #   - Use UPPER_SNAKE_CASE (e.g. WEATHER_API_KEY)
    #   - Align with platform provider's credential_mapping for automatic mapping
    #     e.g.: TWITTER_ACCESS_TOKEN, GITHUB_TOKEN, GOOGLE_ACCESS_TOKEN
    #   - Custom services: SERVICE_NAME + field type
    #
    # sensitive=True credentials are displayed as password fields in UI, not echoed.
    "credentials": [
        {
            "name": "WEATHER_API_KEY",
            "display_name": "OpenWeatherMap API Key",
            "description": "API Key obtained from https://openweathermap.org/api",
            "required": True,
            "sensitive": True,
        },
        {
            "name": "WEATHER_UNITS",
            "display_name": "Temperature Units",
            "description": "Temperature unit preference: metric (Celsius) / imperial (Fahrenheit) / standard (Kelvin)",
            "required": False,
            "sensitive": False,
            "default": "metric",
        },
    ],
    "tools": [
        {
            "name": "get_weather",
            "description": "Query current weather for a specified city",
            "parameters": [
                {
                    "name": "city",
                    "type": "string",
                    "description": "City name (in English), e.g. Beijing, Tokyo, London",
                    "required": True,
                },
            ],
        },
        {
            "name": "get_forecast",
            "description": "Query weather forecast for a specified city",
            "parameters": [
                {
                    "name": "city",
                    "type": "string",
                    "description": "City name (in English)",
                    "required": True,
                },
                {
                    "name": "days",
                    "type": "integer",
                    "description": "Number of forecast days (1-5)",
                    "required": False,
                    "default": 3,
                },
            ],
        },
    ],
    "runtime": {
        "type": "uv",
        "min_version": "0.1.0",
    },
}


# ─── Tool Implementation ─────────────────────────────────────────────


def tool_get_weather(city: str, *, credentials: dict | None = None) -> dict:
    """Query current weather for a specified city

    Credential resolution priority:
    1. context.credentials (platform authorization / plugin-level, Agent-injected)
    2. Environment variables (local development fallback)

    In a real implementation, this would use the API Key from credentials to call an external weather API.
    This example returns simulated data to demonstrate the credential injection flow.
    """
    creds = credentials or {}

    # Best practice: read from context.credentials first, fall back to env vars
    api_key = creds.get("WEATHER_API_KEY") or os.environ.get("WEATHER_API_KEY")
    units = creds.get("WEATHER_UNITS") or os.environ.get("WEATHER_UNITS", "metric")

    if not api_key:
        return {
            "error": "WEATHER_API_KEY not configured",
            "hint": (
                "Configuration options (choose one):\n"
                "  1. Platform authorization: /settings/authorizations page\n"
                "  2. Plugin-level credentials: Anna Admin → Plugin Settings → Credentials\n"
                "  3. Local development: WEATHER_API_KEY=xxx python credential_plugin.py"
            ),
        }

    # ─── Actual API call example (commented out) ───
    # import urllib.request
    # url = (
    #     f"https://api.openweathermap.org/data/2.5/weather"
    #     f"?q={city}&appid={api_key}&units={units}"
    # )
    # resp = urllib.request.urlopen(url)
    # data = json.loads(resp.read())
    # ────────────────────────────

    # Simulated data (for demonstration)
    unit_symbol = {"metric": "°C", "imperial": "°F", "standard": "K"}.get(units, "°C")
    return {
        "city": city,
        "temperature": f"22{unit_symbol}",
        "humidity": "65%",
        "description": "partly cloudy",
        "wind_speed": "3.5 m/s",
        "api_key_configured": True,
        "api_key_preview": f"{api_key[:4]}...{api_key[-4:]}" if len(api_key) > 8 else "***",
        "units": units,
        "_note": "This is simulated data for demonstration purposes",
    }


def tool_get_forecast(
    city: str, days: int = 3, *, credentials: dict | None = None
) -> dict:
    """Query weather forecast for a specified city"""
    creds = credentials or {}
    api_key = creds.get("WEATHER_API_KEY") or os.environ.get("WEATHER_API_KEY")
    units = creds.get("WEATHER_UNITS") or os.environ.get("WEATHER_UNITS", "metric")

    if not api_key:
        return {
            "error": "WEATHER_API_KEY not configured",
            "hint": (
                "Configuration options (choose one):\n"
                "  1. Platform authorization: /settings/authorizations page\n"
                "  2. Plugin-level credentials: Anna Admin → Plugin Settings → Credentials\n"
                "  3. Local development: WEATHER_API_KEY=xxx python credential_plugin.py"
            ),
        }

    days = max(1, min(5, days))
    unit_symbol = {"metric": "°C", "imperial": "°F", "standard": "K"}.get(units, "°C")

    # Simulated forecast data
    forecast = []
    for i in range(days):
        temp = 20 + i * 2
        forecast.append({
            "day": i + 1,
            "temperature_high": f"{temp + 5}{unit_symbol}",
            "temperature_low": f"{temp - 3}{unit_symbol}",
            "description": ["sunny", "cloudy", "rain", "thunderstorm", "clear"][i % 5],
        })

    return {
        "city": city,
        "days": days,
        "forecast": forecast,
        "api_key_configured": True,
        "_note": "This is simulated data for demonstration purposes",
    }


# ─── Tool Dispatch Table ─────────────────────────────────────────────

TOOL_DISPATCH = {
    "get_weather": tool_get_weather,
    "get_forecast": tool_get_forecast,
}


# ─── JSON-RPC Handling ───────────────────────────────────────────────


def make_response(id, result=None, error=None):
    """Construct a JSON-RPC 2.0 response"""
    resp = {"jsonrpc": "2.0", "id": id}
    if error is not None:
        resp["error"] = error
    else:
        resp["result"] = result
    return resp


def handle_describe(request_id):
    """Handle describe request — return the tool manifest (including credentials declaration)"""
    return make_response(request_id, result=MANIFEST)


def handle_invoke(request_id, params):
    """Handle invoke request — execute a tool call

    Credentials are injected via params.context.credentials, format:
    {
        "tool": "get_weather",
        "arguments": {"city": "Beijing"},
        "context": {
            "credentials": {
                "WEATHER_API_KEY": "ak_xxxxx",
                "WEATHER_UNITS": "metric"
            }
        }
    }

    Note: Credentials are fetched and decrypted by the Agent from Nexus before injection.
    The LLM cannot see credential values, and the plugin does not need to manage credential storage.
    """
    tool_name = params.get("tool")
    arguments = params.get("arguments", {})

    # Extract credentials from context (injected by Agent)
    context = params.get("context", {})
    credentials = context.get("credentials")

    if not tool_name:
        return make_response(
            request_id,
            error={"code": -32602, "message": "Missing 'tool' in params"},
        )

    fn = TOOL_DISPATCH.get(tool_name)
    if fn is None:
        return make_response(
            request_id,
            error={
                "code": -32601,
                "message": f"Unknown tool: {tool_name}",
                "data": {"available_tools": list(TOOL_DISPATCH.keys())},
            },
        )

    try:
        # Pass credentials as a keyword argument to the tool function
        result = fn(**arguments, credentials=credentials)
        return make_response(
            request_id,
            result={"success": True, "data": result, "tool": tool_name},
        )
    except TypeError as e:
        return make_response(
            request_id,
            error={"code": -32602, "message": f"Invalid parameters: {e}"},
        )
    except Exception as e:
        return make_response(
            request_id,
            error={"code": -32603, "message": f"Tool execution failed: {e}"},
        )


def handle_health(request_id):
    """Handle health request — health check"""
    return make_response(
        request_id,
        result={
            "status": "healthy",
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "version": MANIFEST["version"],
            "tools_count": len(MANIFEST["tools"]),
            "credentials_declared": len(MANIFEST.get("credentials", [])),
        },
    )


def handle_request(line: str) -> str:
    """Parse and handle a single JSON-RPC request"""
    try:
        request = json.loads(line)
    except json.JSONDecodeError:
        return json.dumps(
            make_response(None, error={"code": -32700, "message": "Parse error"})
        )

    request_id = request.get("id")
    method = request.get("method")
    params = request.get("params", {})

    if method == "describe":
        response = handle_describe(request_id)
    elif method == "invoke":
        response = handle_invoke(request_id, params)
    elif method == "health":
        response = handle_health(request_id)
    else:
        response = make_response(
            request_id,
            error={"code": -32601, "message": f"Method not found: {method}"},
        )

    return json.dumps(response)


# ─── Main Loop (stdio JSON-RPC Service) ──────────────────────────────


def main():
    """Main entry point: reads JSON-RPC requests line by line from stdin, returns responses via stdout.

    Important: All log output goes to stderr to avoid interfering with protocol communication.
    """
    print("🔌 Weather credential plugin started", file=sys.stderr)
    print(f"   Tools: {list(TOOL_DISPATCH.keys())}", file=sys.stderr)
    print(
        f"   Credentials required: "
        f"{[c['name'] for c in MANIFEST.get('credentials', [])]}",
        file=sys.stderr,
    )

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        print(f"← {line}", file=sys.stderr)

        response = handle_request(line)

        # Response via stdout — must flush to avoid buffer blocking
        print(response, flush=True)

        print(f"→ {response}", file=sys.stderr)


if __name__ == "__main__":
    main()
